File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        int low=0, high=0, n=s.length(), len=0;

        Map<Character, Integer> map = new HashMap<>();

        while(high<n){
            map.put(s.charAt(high), map.getOrDefault(s.charAt(high), 0)+1);
            if(map.get(s.charAt(high))==1){
                len = Math.max(len, high-low+1);
                high++;
            }
            else{
                while(map.get(s.charAt(high))!=1){
                    map.put(s.charAt(low), map.get(s.charAt(low))-1);
                    if(map.get(s.charAt(low))==0){
                        map.remove(s.charAt(low));
                    }
                    low++;
                }
                high++;
            }
        }
        if(len!=0){
            return len;
        }
        else{
            return 0;
        }

    }
}

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.py
from collections import defaultdict
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        n=len(s)
        low=0
        high=0
        res=0
        charMap=defaultdict(int)
        while high<n:
            charMap[s[high]]+=1
            if charMap[s[high]] <= 1:
                res = max(res, high-low+1)
                high+=1
            else:   #duplicate found
                while charMap[s[high]] > 1:
                    charMap[s[low]] -=1
                    if charMap[s[low]] == 0:
                        del charMap[s[low]]
                    low+=1
                high+=1
        return res
